// victory-voronoi/lib/victory-voronoi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VictoryVoronoi = void 0;
var _react = _interopRequireDefault(require("react"));
var _victoryCore = require("victory-core");
var _voronoi = require("./voronoi");
var _helperMethods = require("./helper-methods");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const fallbackProps = {
  width: 450,
  height: 300,
  padding: 50
};

// eslint-disable-next-line @typescript-eslint/no-empty-interface

class VictoryVoronoiBase extends _react.default.Component {
  static animationWhitelist = ["data", "domain", "height", "padding", "samples", "size", "style", "width"];
  static displayName = "VictoryVoronoi";
  static role = "voronoi";
  static defaultTransitions = _victoryCore.DefaultTransitions.discreteTransitions();
  static defaultProps = {
    containerComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryContainer, null),
    dataComponent: /*#__PURE__*/_react.default.createElement(_voronoi.Voronoi, null),
    labelComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryLabel, null),
    groupComponent: /*#__PURE__*/_react.default.createElement("g", {
      role: "presentation"
    }),
    samples: 50,
    sortOrder: "ascending",
    standalone: true,
    theme: _victoryCore.VictoryTheme.grayscale
  };
  static getDomain = _victoryCore.Domain.getDomain;
  static getData = _victoryCore.Data.getData;
  static getBaseProps(props) {
    return (0, _helperMethods.getBaseProps)(props, fallbackProps);
  }
  static expectedComponents = ["dataComponent", "labelComponent", "groupComponent", "containerComponent"];

  // Overridden in native versions
  shouldAnimate() {
    return !!this.props.animate;
  }
  render() {
    const {
      animationWhitelist,
      role
    } = VictoryVoronoi;
    const props = _victoryCore.Helpers.modifyProps(this.props, fallbackProps, role);
    if (this.shouldAnimate()) {
      return this.animateComponent(props, animationWhitelist);
    }
    const children = this.renderData(props);
    const component = props.standalone ? this.renderContainer(props.containerComponent, children) : children;
    return _victoryCore.UserProps.withSafeUserProps(component, props);
  }
}
const VictoryVoronoi = exports.VictoryVoronoi = (0, _victoryCore.addEvents)(VictoryVoronoiBase);
