// victory-errorbar/lib/victory-errorbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VictoryErrorBar = void 0;
var _react = _interopRequireDefault(require("react"));
var _victoryCore = require("victory-core");
var _errorBar = require("./error-bar");
var _helperMethods = require("./helper-methods");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const fallbackProps = {
  width: 450,
  height: 300,
  padding: 50
};
const defaultData = [{
  x: 1,
  y: 1,
  errorX: 0.1,
  errorY: 0.1
}, {
  x: 2,
  y: 2,
  errorX: 0.2,
  errorY: 0.2
}, {
  x: 3,
  y: 3,
  errorX: 0.3,
  errorY: 0.3
}, {
  x: 4,
  y: 4,
  errorX: 0.4,
  errorY: 0.4
}];

// eslint-disable-next-line @typescript-eslint/no-empty-interface

class VictoryErrorBarBase extends _react.default.Component {
  static animationWhitelist = ["data", "domain", "height", "padding", "samples", "style", "width", "errorX", "errorY", "borderWidth"];
  static displayName = "VictoryErrorBar";
  static role = "errorbar";
  static defaultTransitions = _victoryCore.DefaultTransitions.discreteTransitions();
  static defaultProps = {
    containerComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryContainer, null),
    data: defaultData,
    dataComponent: /*#__PURE__*/_react.default.createElement(_errorBar.ErrorBar, null),
    labelComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.VictoryLabel, null),
    groupComponent: /*#__PURE__*/_react.default.createElement("g", {
      role: "presentation"
    }),
    samples: 50,
    sortOrder: "ascending",
    standalone: true,
    theme: _victoryCore.VictoryTheme.grayscale
  };
  static getDomain(props, axis) {
    return (0, _helperMethods.getDomain)(props, axis);
  }
  static getData(props) {
    return (0, _helperMethods.getData)(props);
  }
  static getBaseProps(props) {
    return (0, _helperMethods.getBaseProps)(props, fallbackProps);
  }
  static expectedComponents = ["dataComponent", "labelComponent", "groupComponent", "containerComponent"];

  // Overridden in native versions
  shouldAnimate() {
    return !!this.props.animate;
  }
  render() {
    const {
      animationWhitelist,
      role
    } = VictoryErrorBar;
    const props = _victoryCore.Helpers.modifyProps(this.props, fallbackProps, role);
    if (this.shouldAnimate()) {
      return this.animateComponent(props, animationWhitelist);
    }
    const children = this.renderData(props);
    const component = props.standalone ? this.renderContainer(props.containerComponent, children) : children;
    return _victoryCore.UserProps.withSafeUserProps(component, props);
  }
}
const VictoryErrorBar = exports.VictoryErrorBar = (0, _victoryCore.addEvents)(VictoryErrorBarBase);
